agent/consul/multilimiter/multilimiter.go

```diff
@@ -0,0 +1,148 @@
+package multilimiter
+
+import (
+	"context"
+	radix "github.com/hashicorp/go-immutable-radix"
+	"golang.org/x/time/rate"
+	"sync/atomic"
+	"time"
+)
+
+var _ RateLimiter = &MultiLimiter{}
+
+// RateLimiter is the interface implemented by MultiLimiter
+type RateLimiter interface {
+	Run(ctx context.Context)
+	Allow(entity LimitedEntity) bool
+	UpdateConfig(c Config)
+}
+
+// MultiLimiter implement RateLimiter interface and represent a set of rate limiters
+// specific to different LimitedEntities and queried by a LimitedEntities.Key()
+type MultiLimiter struct {
+	limiters *atomic.Pointer[radix.Tree]
+	config   *atomic.Pointer[Config]
+}
+
+// LimitedEntity is an interface used by MultiLimiter.Allow to determine
+// which rate limiter to use to allow the request
+type LimitedEntity interface {
+	Key() []byte
+}
+
+// Limiter define a limiter to be part of the MultiLimiter structure
+type Limiter struct {
+	limiter    *rate.Limiter
+	lastAccess atomic.Int64
+	config     atomic.Pointer[LimiterConfig]
```

Comment:
I've just realised that `config` applies to **all** limiters/keys. Does that mean we'll have separate `MultiLimiter`s for global read, global write, ip-based read and ip-based write?

This seems counter to the following diagram from the RFC:

![image](https://user-images.githubusercontent.com/580803/203524527-e731623b-bd1b-4dac-b5d4-44de879f99af.png)

Reply:
Yes, That part of the design need to change IMO. I will update the RFC. Because the first Key (operation.read/write) is a finite number I think it's overkilled to embedded into the `MultiLimiter` and we can instead have a `map[string]MultiLimiter` created when bootstrapping or updating the config. What you think about that?

Reply:
I guess I was imaging that the radix tree was storing all limiters and that the keys would be based on the properties defined in the diagram.  Those properties being (rough names used here):
- Operation Category - GLOBAL, KVS, CATALOG, etc. (transaction type currently part of brown "column" in the diagram)
- Operation Type - WRITE,READ (transaction type currently part of brown "column" in the diagram)
- Limit Type - server/global/all, IP, tenant, service (this prevents name/key collisions.   corresponds to the phases of the implementation)
- Scope - global, 10.1.1.11, tenant-a, checkout-service, etc (this is all the possible values across all phases for the yellow column in the diagram)

So naively when I looked at LimitedEntity I imagined keys would look like this:
 ```
/<limit type>/<scope>/<category>/<operation type>
EX:
/server/all/global/read- phase 1
/ip/10.1.1.11/global/write/ - phase 2
/ip/10.1.1.11/catalog/write/ - phase 2
/tenant/tenant-a/global/write/ - phase 3
/service/checkout-service/global/write/ - phase 4
```

The above was really based on the LimitedEntity and how might iterate on it through phases, but possibly the above has storage concerns or other things.

Reply:
I also wonder if having `map[string]MultiLimiter` would cause us to use locking and atomic operations for this as well?

Reply:
Yeah, I had assumed the same 😅

Without storing configuration on the tree leaf nodes, we'd need to track the identity/tenant-specific configuration in some other structure - which seems like a shame when you've put so much effort into making this one efficient!